### PR TITLE
Issue/4315 scenario1

### LINKED ISSUE #4315: [E2E] Scenario 1: Environment Scenarios
# Scenario 1 : Environment
## 1.1 Cancel creation
- click on card 'Create new environment'
- fill in Project Name
- fill in Name
- fill in Description
- fill in Repository
- fill in Branch
- add Icon
- Cancel form (Expect alert if already implemented and confirm alert)
- Expect to go back to home page
- Expect no new Environment created

## 1.2 Submit form
- click on card 'Create new environment'
- fill in Project Name
- OMIT to fill in Name
- fill in Description
- fill in Repository
- fill in Branch
- add Icon
- Try to submit form
- Expect submit button to be disabled
- fill in Name
- Submit form
- Expect to go the Service catalog from the created env
- Expect new Environment Service catalog to be empty.
- Go back to home page
- Expect to find the card from the created env.

## 1.3 Delete Created Environment
- click on card 'Create new environment'
- fill in Project Name
- fill in Name
- Submit form
- Expect to go to Service Catalog from newly created env.
- Go to Env Settings
- Expect name to match
- click delete environment
- Fill in modal with env name
- submit modal
- Expect to be redirected to home page
- Expect to not find the previously created Env in the cards.

## 1.4 Edit Created Environment
- click on card 'Create new environment'
- fill in Project Name
- fill in Name
- Submit form
- Expect to go to Service Catalog from newly created env.
- Go to Env Settings
- Expect name to match
*For Each field*
    - Click on pen field
    - change value
    - submit
    - expect new value in settings
- Delete instance to cleanup
- Expect to be redirected to home page

## 1.5 Clear Environment 
- Select test environment
- expect a to have items in the Service Catalog
- go to settings 
- Click on Clear environment
- Cancel modal
- Go to Service Catalog
- Expect nothing to have changed 
- Go back to settings
- Click on Clear environment
- Confirm modal by writing name of environment and clicking on submit button
- Go back to Service Catalog
- Expect it to be empty.
- Click on Update Service Catalog
- Await compilation
- Expect services to be back.
- Go back to home

## 1.6 Edit Configuration
- click on card 'Create new environment'
- fill in Project Name
- fill in Name
- Submit form
- Expect to go to Service Catalog from newly created env.
- Go to Env Settings
- Expect name to match
- Go to Configuration tab
*For Each field*
    - change value
    - submit
    - expect new value in settings
- Go to Environment tab
- Delete instance to cleanup
- Expect to be redirected to home page

### PR DESCRIPTION
**Pull request by matborowczyk**
# Description

* Short description here *

closes #4315 

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [ ] Attached issue to pull request
- [ ] Changelog entry
- [ ] Code is clear and sufficiently documented
- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Correct, in line with design
~~- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )~~

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

### CODE REVIEW

File: cypress/e2e/environment.cy.js [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+    );
+    createEnv(1);
+    cy.get("button").contains("Cancel").click();
+    cy.url().should("eq", Cypress.config().baseUrl + "/console/");
+  });
+
+  it("1.2 Create ne environment", function () {
+    //fill the form and submit
+    cy.visit("/console/environment/create");
+    createEnv(2, false);
```
Comment (LauraStordeur) [inline cypress/e2e/environment.cy.js:173]: Don't use nameless arguments, it's good practice for more readable code :)

File: cypress/e2e/environment.cy.js [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+
+  it("1.3 delete an environment", function () {
+    //Fill The form and submit
+    cy.visit("/console/environment/create");
+    createEnv(3, true, true);
+    cy.get("button").contains("Submit").click();
+    cy.url().should("contain", "/console/lsm/catalog?env=");
+    //Go to settings
+    cy.get('[aria-label="Settings actions"]').click();
+    cy.url().should("contain", "/console/settings?env=");
```
Comment (LauraStordeur) [inline cypress/e2e/environment.cy.js:196]: path in object

File: cypress/e2e/environment.cy.js [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+  });
+
+  it("1.3 delete an environment", function () {
+    //Fill The form and submit
+    cy.visit("/console/environment/create");
+    createEnv(3, true, true);
+    cy.get("button").contains("Submit").click();
+    cy.url().should("contain", "/console/lsm/catalog?env=");
+    //Go to settings
+    cy.get('[aria-label="Settings actions"]').click();
```
Comment (LauraStordeur) [inline cypress/e2e/environment.cy.js:195]: Go to settings is being reused in a few places too, handy to have it a as a method

File: cypress/e2e/environment.cy.js [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+    cy.url().should("contain", "/console/settings?env=");
+    cy.get('[aria-label="Name-value"]').should("contain", testName(3));
+    deleteEnv(3);
+  });
+
+  it("1.4 Edit created environment", function () {
+    //Fill The form and submit
+    cy.visit("/console/environment/create");
```
Comment (LauraStordeur) [inline cypress/e2e/environment.cy.js:203]: path in object :)

File: cypress/e2e/environment.cy.js [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+    createEnv(4, true, true);
```
Comment (LauraStordeur) [inline cypress/e2e/environment.cy.js:204]: nameless args

File: cypress/e2e/environment.cy.js [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+          break;
+        case "icon":
+          cy.get(iconInput).attachFile(icon);
+          break;
+        case "typeahead":
+          cy.get('[aria-label="Project Name-typeahead"]')
+            .clear()
+            .type("New Value " + parameter.name);
+
+          cy.get('[aria-label="Project Name-select-input"]').then(($list) => {
```
Comment (LauraStordeur) [inline cypress/e2e/environment.cy.js:230]: why the dollar sign in your variable?

File: cypress/e2e/environment.cy.js [hunk 7]
```diff
@@ -0,0 +1,389 @@
...
+      .contains("I understand the consequences, clear this environment")
+      .click();
+    cy.visit("/console/");
+    cy.get('[aria-label="Environment card"]').contains("test").click();
+    cy.get('[aria-label="ServiceCatalog-Empty"]').should("to.be.visible");
+    //Update service catalog to restore instances
+    cy.get("button").contains("Update Service Catalog").click();
+    cy.get("button").contains("Yes").click();
+    cy.intercept(
+      "/api/v2/environment/a5f9f7f2-5f43-4fad-a16d-f71f127f0e91?details=fals"
```
Comment (LauraStordeur) [inline cypress/e2e/environment.cy.js:317]: This intercept with hardcoded id might cause problems, also I think you miss an 'e' at false

File: cypress/e2e/environment.cy.js [hunk 8]
```diff
@@ -0,0 +1,389 @@
...
+    cy.get("button").contains("Clear environment").click();
+    cy.get('[aria-label="clear environment check"]').type("test");
+    cy.get("button")
+      .contains("I understand the consequences, clear this environment")
+      .click();
+    cy.visit("/console/");
+    cy.get('[aria-label="Environment card"]').contains("test").click();
+    cy.get('[aria-label="ServiceCatalog-Empty"]').should("to.be.visible");
+    //Update service catalog to restore instances
+    cy.get("button").contains("Update Service Catalog").click();
```
Comment (LauraStordeur) [inline cypress/e2e/environment.cy.js:314]: words dict

File: cypress/e2e/environment.cy.js [hunk 9]
```diff
@@ -0,0 +1,389 @@
...
+    cy.get("button").contains("Yes").click();
+    cy.intercept(
+      "/api/v2/environment/a5f9f7f2-5f43-4fad-a16d-f71f127f0e91?details=fals"
+    ).as("catalogUpdate");
+    cy.wait("@catalogUpdate")
+      .its("response.body")
+      .then(() => {
+        cy.get('[aria-label="ServiceCatalog-Success"]').should("to.be.visible");
```
Comment (LauraStordeur) [inline cypress/e2e/environment.cy.js:322]: that's the toaster, is I'm not mistaken, but we also need to assert that the services are back in the list :)

File: cypress/e2e/environment.cy.js [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+      });
+  });
+
+  it("1.6 Edit environment configuration", function () {
+    cy.visit("/console/environment/create");
+    createEnv(6, true, true);
```
Comment (LauraStordeur) [inline cypress/e2e/environment.cy.js:328]: nameless args

File: cypress/e2e/environment.cy.js [hunk 11]
```diff
@@ -0,0 +1,389 @@
...
+    cy.get("button").contains("Submit").click();
+    cy.url().should("contain", "/console/lsm/catalog?env=");
```
Comment (LauraStordeur) [inline cypress/e2e/environment.cy.js:330]: url to object

File: cypress/e2e/environment.cy.js [hunk 12]
```diff
@@ -0,0 +1,389 @@
...
+
```
Comment (LauraStordeur) [inline cypress/e2e/environment.cy.js:331]: you added a comment in previous steps ;) would be nice to keep this :)

File: cypress/e2e/environment.cy.js [hunk 13]
```diff
@@ -0,0 +1,389 @@
...
+    cy.get('[aria-label="Settings actions"]').click();
+    cy.get('[aria-label="Name-value"]').should("eq", testName(6));
+    cy.get('[aria-label="Settings actions"]').click();
+    cy.get("button").contains("Configuration");
+    cy.wrap(configurationTypes).each((config) => {
+      switch (config.type) {
+        case "text":
+          cy.get("Row-" + config.name + " > td > input")
```
Comment (LauraStordeur) [inline cypress/e2e/environment.cy.js:339]: you created an array of objects why not add the path for each config there too?

File: cypress/e2e/environment.cy.js [hunk 14]
```diff
@@ -0,0 +1,389 @@
...
+
+          break;
+        case "select":
+          cy.get("Row-" + config.name + " > td > input").click();
+          cy.get("Row-" + config.name + " > td > .pf-c-select__menu-item")
+            .contains(config.newValue)
+            .click();
+
+          break;
+        case "selecMap":
```
Comment (LauraStordeur) [inline cypress/e2e/environment.cy.js:354]: selectMap, I suppose? Seems like you miss a 't' in the word

**Review by LauraStordeur — COMMENTED** [review]
I didn't comment on every single row, sorry for the lots of them already. I know you're not finished yet, but I thought it might be useful to already go through your code. 

Most comments relate to 'reusablity' of the code with the idea in mind that we will have lot's of tests to write. 

I also have a strong opinion about 'magic' strings,  may it be in tests or in the code itself from the app. This might need to be discussed to figure out what's the best rule for it :)

File: cypress/e2e/environment.cy.js [hunk 15]
```diff
@@ -0,0 +1,389 @@
...
+
+  it("1.2 Create ne environment", function () {
+    //fill the form and submit
+    cy.visit("/console/environment/create");
+    createEnv(2, false);
+    cy.get("button").contains("Submit").should("be.disabled");
+    cy.get(nameInput).type(testName(2));
+    cy.get("button").contains("Submit").should("not.be.disabled");
+    cy.get("button").contains("Submit").click();
+    cy.url().should("contain", "/console/lsm/catalog?env=");
```
[the remark this answers is not in the document]
Comment (bartv) [reply cypress/e2e/environment.cy.js:178]: This really depends on how much it will be used. Factoring this all out in constants is not going to make the code more readable. Especially in test cases I am actually against this. When this changes, you can also do a search and replace.

[the remark this answers is not in the document]
Comment (LauraStordeur) [reply cypress/e2e/environment.cy.js:178]: It actually is being reused, and I believe having an object like so
```javascript
const PATHS = {
     create_env: "/console/lsm/catalog?env=",
     settings_env: "....",
```
would be useful

File: cypress/e2e/environment.cy.js [hunk 16]
```diff
@@ -0,0 +1,389 @@
...
+    createEnv(1);
+    cy.get("button").contains("Cancel").click();
+    cy.url().should("eq", Cypress.config().baseUrl + "/console/");
+  });
+
+  it("1.2 Create ne environment", function () {
+    //fill the form and submit
+    cy.visit("/console/environment/create");
+    createEnv(2, false);
+    cy.get("button").contains("Submit").should("be.disabled");
```
[the remark this answers is not in the document]
Comment (wouterdb) [reply cypress/e2e/environment.cy.js:174]: I strongly oppose this in this case.

There are reasons to factor out common code, but I don't see which ones apply here. 

For me, in testcases, the rule is that it should read as a story. It should, at every point, be clear what is happening.
At this point, this is very readable. It relates well to what is visible. 

By factoring out this trivial code, it becomes harder to read, because we replace something concrete with something abstract, that may not be as clear as the current code. 

At the same time, it reduces locality: to modify this assertion, the modification has to be done somewhere else. 
It also tangles the testcases: when we update the assert, we have to check that it does not affect any other place it is used. 

In general, for me, the idea is not to factor out common code, but strong abstractions. 
(I can explain this more in detail in a meeting)

File: cypress/e2e/environment.cy.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (wouterdb) [reply cypress/e2e/environment.cy.js:178]: This is similar to the previous case, 

Neither the argument `The url is or should ba a constant, can be extracted` nor `It actually is being reused,` 
carry any weight with me, because I don't see any intrinsic value in either. 

However, you do neatly demonstrate that there is an abstraction that can be extracted: the overview of paths. 

So here the question is harder for me to answer, because there are some delicate balances:
- reduced locality, increased tangling
- potential increase in readability (by naming URL's), potential decreased readability (by hiding urls)
- very subtle change in what you test: the url is exactly this Vs what is expected in general (would matter if we test for URL compatibility over versions, but we don't so it doesn't really matter)  
- easier to change routing (ridiculous argument, as I expect all hell to break loose anyways and this will be a drop in the sea)

File: cypress/e2e/environment.cy.js [hunk 17]
```diff
@@ -0,0 +1,389 @@
...
+    cy.get(nameInput).type(testName(2));
+    cy.get("button").contains("Submit").should("not.be.disabled");
+    cy.get("button").contains("Submit").click();
+    cy.url().should("contain", "/console/lsm/catalog?env=");
+    //go back to gome and check if env is visible
+    cy.get(".pf-c-breadcrumb__item").first().click();
+    cy.url().should("eq", Cypress.config().baseUrl + "/console/");
+    cy.get('[aria-label="Environment card"]').should(
```
[the remark this answers is not in the document]
Comment (wouterdb) [reply cypress/e2e/environment.cy.js:182]: what abstraction would this bring?

File: cypress/e2e/environment.cy.js (hunk 8 above)
Comment (wouterdb) [reply cypress/e2e/environment.cy.js:314]: i don't understand this?

Comment (LauraStordeur) [reply cypress/e2e/environment.cy.js:314]: This text is a printed value on the screen and is available in the `words` dict method used  for these values :)

File: cypress/e2e/environment.cy.js (hunk 9 above)
Comment (matborowczyk) [reply cypress/e2e/environment.cy.js:322]: that's label corresponds to the component that holds these services, if you think that I should also assert toast, I'll add it

File: cypress/e2e/environment.cy.js (hunk 12 above)
Comment (matborowczyk) [reply cypress/e2e/environment.cy.js:331]: That's the plan :D I left it when all tests will be properly set up

File: cypress/e2e/environment.cy.js (hunk 13 above)
Comment (matborowczyk) [reply cypress/e2e/environment.cy.js:339]: Fair point, didn't think of that

File: cypress/e2e/environment.cy.js (hunk 14 above)
Comment (matborowczyk) [reply cypress/e2e/environment.cy.js:354]: Yes, That's the part that is still in progress, but good point, less chance that I'll forget about that :D

File: cypress/e2e/environment.cy.js (hunk 9 above)
Comment (LauraStordeur) [reply cypress/e2e/environment.cy.js:322]: Maybe it would be good to check if the length of the list of services is the same as before clearing the environment 🤔

File: cypress/e2e/environment.cy.js (hunk 6 above)
Comment (matborowczyk) [reply cypress/e2e/environment.cy.js:230]: That's the convention cypress used in the docs to mark html components, at least in snippets they have, I followed, but I can change it, as I don't mind :)

Comment (LauraStordeur) [reply cypress/e2e/environment.cy.js:230]: I don't mind it, I just never saw that. That's why I was wondering what it meant

File: cypress/e2e/environment.cy.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (LauraStordeur) [reply cypress/e2e/environment.cy.js:174]: As discussed we will only extract reussable sections when they appear to be usefull in other test scenarios.

File: cypress/e2e/environment.cy.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (LauraStordeur) [reply cypress/e2e/environment.cy.js:178]: As agreed on in the meeting, this is a logic ensemble of values, but it doesn't bring that much benefit to store them in objects. Both solutions are fine for me. @matborowczyk you can resolve the thread and specify whichever solution you decided to pick, I'll keep in sync then with the tests I'm writing ;)

File: cypress/e2e/environment.cy.js (hunk 6 above)
Comment (LauraStordeur) [reply cypress/e2e/environment.cy.js:230]: Agreed to follow the code guidelines from cypress and use the dollar sign to specify HTML elements and components.

File: cypress/e2e/environment.cy.js [hunk 18]
```diff
@@ -0,0 +1,389 @@
...
+        }
+      }
+    });
+    deleteEnv(4);
+  });
+
+  it("1.5 Clear environment", function () {
+    //Fill The form and submit
+    cy.visit("/console/");
+    cy.get('[aria-label="Environment card"]').contains("test").click();
```
Comment (LauraStordeur) [inline cypress/e2e/environment.cy.js:284]: I noticed that it will pick any environment that 'contains' the word test, the idea was to use the frontend-lsm one, in the test I'm writing I had to use the project label to be sure it picked the right one.

File: cypress/e2e/environment.cy.js [hunk 19]
```diff
@@ -0,0 +1,389 @@
...
+    cy.get('[aria-label="Settings actions"]').click();
+    cy.get('[aria-label="Name-value"]').should("eq", testName(6));
+    cy.get('[aria-label="Settings actions"]').click();
+    cy.get("button").contains("Configuration");
+    cy.wrap(configurationTypes).each((config) => {
+      switch (config.type) {
+        case "text":
+          cy.get("Row-" + config.name + " > td > input")
+            .clear()
+            .tpye(config.newValue);
```
Comment (LauraStordeur) [inline cypress/e2e/environment.cy.js:341]: type*

File: cypress.config.cjs [hunk 20]
```diff
@@ -8,7 +8,7 @@ module.exports = defineConfig({
     mochaFile: "cypress/reports/junit/test-report-[hash].xml",
   },
   e2e: {
-    baseUrl: "http://localhost:8888",
+    baseUrl: "http://localhost:9000",
```
Comment (LauraStordeur) [inline cypress.config.cjs:11]: why did you change it back?

File: cypress/e2e/environment.cy.js (hunk 18 above)
Comment (matborowczyk) [reply cypress/e2e/environment.cy.js]: You are right, I changed it

File: cypress/e2e/environment.cy.js [hunk 21]
```diff
@@ -0,0 +1,483 @@
...
+              })
+              .clear()
+              .type(config.newValue);
+          });
+          break;
+      }
+      //assert if the warning sign exist, which means that value in the input is new but not saved, then save, and expect warning sign to disapear
+      cy.wait(100);
+      cy.get('[aria-label="Warning"]').should("exist");
+      cy.get(config.row).within(() => {
```
Comment (LauraStordeur) [inline cypress/e2e/environment.cy.js:442]: https://docs.cypress.io/api/commands/within#Syntax

File: cypress/e2e/environment.cy.js [hunk 22]
```diff
@@ -0,0 +1,483 @@
...
+        cy.get('[aria-label="SaveAction"]').click();
+      });
+      cy.wait(100);
+      cy.get('[aria-label="Warning"]').should("not.exist");
+      //go through every configuration and expect new value there
+      switch (config.inputType) {
+        case "textNumber" | "text":
+          cy.get(config.row).within(() => {
```
Comment (LauraStordeur) [inline cypress/e2e/environment.cy.js:450]: https://docs.cypress.io/api/commands/within#Syntax

File: cypress/e2e/environment.cy.js [hunk 23]
```diff
@@ -0,0 +1,483 @@
...
+            cy.get(".pf-c-form-control").should("have.text", config.newValue);
+          });
+          break;
+        case "select":
+          cy.get(config.row).within(() => {
```
Comment (LauraStordeur) [inline cypress/e2e/environment.cy.js:455]: https://docs.cypress.io/api/commands/within#Syntax

File: cypress/e2e/environment.cy.js [hunk 24]
```diff
@@ -0,0 +1,483 @@
...
+            cy.get('[aria-label="editEntryValue"]').should(
+              "have.value",
+              config.newValue
+            );
+          });
+          break;
+      }
+    });
+    //re-enable to delete env
+    cy.get('[aria-label="Row-protected_environment"]').within(() => {
```
Comment (LauraStordeur) [inline cypress/e2e/environment.cy.js:471]: https://docs.cypress.io/api/commands/within#Syntax

File: cypress/e2e/environment.cy.js [hunk 25]
```diff
@@ -0,0 +1,483 @@
...
+      cy.get(".pf-c-switch").click();
+    });
+    cy.get('[aria-label="Warning"]').should("exist");
+    cy.get('[aria-label="Row-protected_environment"]').within(() => {
```
Comment (LauraStordeur) [inline cypress/e2e/environment.cy.js:475]: https://docs.cypress.io/api/commands/within#Syntax

File: cypress/e2e/environment.cy.js [hunk 26]
```diff
@@ -0,0 +1,483 @@
...
+          });
+          break;
+        case "select":
+          cy.get(config.row).within(() => {
+            cy.get(".pf-c-form-control").should("have.value", config.newValue);
+          });
+
+          break;
+        case "textmap":
+          cy.get(config.row).within(() => {
```
Comment (LauraStordeur) [inline cypress/e2e/environment.cy.js:461]: https://docs.cypress.io/api/commands/within#Syntax

File: cypress/e2e/environment.cy.js [hunk 27]
```diff
@@ -0,0 +1,483 @@
+const icon = "./cypress/fixtures/test-icon.png";
```
Comment (LauraStordeur) [inline cypress/e2e/environment.cy.js:1]: https://docs.cypress.io/api/commands/fixture#Syntax

I suppose it is best to use the Cypress way of importing ;)

File: cypress/e2e/environment.cy.js (hunk 9 above)
Comment (matborowczyk) [reply cypress/e2e/environment.cy.js]: As I mentioned through the call(but don't want to lost that knowledge so I'll leave it here :) ), that would cause issues, as sometimes we got only part of the services when the status is changed, the rest is added on second call, which is problematic as there is no simple way to assert that in some stable enough way. At least not to my knowledge at the moment. We can come back to this test case in the future.

File: changelogs/unreleased/4315-env-scenarios.yml [hunk 28]
```diff
@@ -0,0 +1,6 @@
+change-type: patch
+description: 'Enviroment e2e test scenarios'
+issue-nr: 4315
+destination-branches:
+- master
+- iso6
```
Comment (LauraStordeur) [inline changelogs/unreleased/4315-env-scenarios.yml:6]: you'll need to add iso5 too, sorry don't remember if I mentioned it, but the bot will not merge it otherwise :D

**Review by LauraStordeur — APPROVED** [review]
